cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc

```diff
@@ -452,6 +474,296 @@ struct Nanosecond {
+      case compute::CalendarUnit::YEAR: {
+        year_month_day ymd(
+            floor<days>(localizer_.template ConvertTimePoint<Duration>(arg)));
+        year y{(static_cast<int32_t>(ymd.year()) / options.multiple) * options.multiple};
+        t = localizer_.ConvertDays(y / jan / 1).time_since_epoch();
+        break;
+      }
+      default:
+        t = Duration{arg};
+    }
+    return static_cast<T>(t.count());
+  }
+
+  Localizer localizer_;
+  RoundTemporalOptions options;
+};
+
+template <typename Duration, typename Localizer>
+struct RoundTemporal {
+  explicit RoundTemporal(const RoundTemporalOptions* options, Localizer&& localizer)
+      : localizer_(std::move(localizer)), options(*options) {}
+
+  template <typename T, typename Arg0>
+  T Call(KernelContext*, Arg0 arg, Status* st) const {
+    Duration t;
+    switch (options.unit) {
+      case compute::CalendarUnit::NANOSECOND:
+        t = RoundTimePoint<Duration, std::chrono::nanoseconds, Localizer>(
+            arg, options.multiple, localizer_, st);
+        break;
+      case compute::CalendarUnit::MICROSECOND:
+        t = RoundTimePoint<Duration, std::chrono::microseconds, Localizer>(
+            arg, options.multiple, localizer_, st);
+        break;
+      case compute::CalendarUnit::MILLISECOND:
+        t = RoundTimePoint<Duration, std::chrono::milliseconds, Localizer>(
+            arg, options.multiple, localizer_, st);
+        break;
+      case compute::CalendarUnit::SECOND:
+        t = RoundTimePoint<Duration, std::chrono::seconds, Localizer>(
+            arg, options.multiple, localizer_, st);
+        break;
+      case compute::CalendarUnit::MINUTE:
+        t = RoundTimePoint<Duration, minutes, Localizer>(arg, options.multiple,
+                                                         localizer_, st);
+        break;
+      case compute::CalendarUnit::HOUR:
+        t = RoundTimePoint<Duration, std::chrono::hours, Localizer>(arg, options.multiple,
+                                                                    localizer_, st);
+        break;
+      case compute::CalendarUnit::DAY:
+        t = RoundTimePoint<Duration, days, Localizer>(arg, options.multiple, localizer_,
+                                                      st);
+        break;
+      case compute::CalendarUnit::WEEK:
+        t = RoundTimePoint<Duration, weeks, Localizer>(arg, options.multiple, localizer_,
+                                                       st);
+        break;
+      case compute::CalendarUnit::MONTH: {
+        auto t0 = localizer_.template ConvertTimePoint<Duration>(arg);
+        year_month_day ymd =
+            GetFlooredYmd<Duration, Localizer>(arg, options.multiple, localizer_);
+
+        auto f = localizer_.ConvertDays(ymd.year() / ymd.month() / 1);
+        ymd += months{options.multiple};
+        auto c = localizer_.ConvertDays(ymd.year() / ymd.month() / 1);
+
+        t = (t0 - f > c - t0) ? c.time_since_epoch() : f.time_since_epoch();
```

Comment:
Just curious, should this be `>=`?

Reply:
I'm not sure how to round half by default. Switching to up (`>=`).